Sporadic test failure with test_timer/test_repeat on Mac OSX and Qt

The following test failure occurs **occasionally** on the Mac OSX build:

```
======================================================================
FAIL: test_repeat (pyface.timer.tests.test_timer.TestTimer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/travis/.edm/envs/pyface-test-3.6-pyside2/lib/python3.6/site-packages/pyface-7.1.0.dev22-py3.6.egg/pyface/timer/tests/test_timer.py", line 425, in test_repeat
    self.assertEqual(handler.count, 4)
AssertionError: 5 != 4
```

For reference, the test that is failing has the Timer's `expire` and `repeat` set to None. So I am not sure why `repeats` can be 0 or less, maybe I was looking at the wrong test?  I am looking at this one:
`pyface.timer.tests.test_timer.TestTimer.test_repeat`

If I just run the test many times, each as a separate process, I could not reproduce it.

But if I rewrite the test like this :

```python
    def test_repeat(self):
        for i in range(40):
            handler = ConditionHandler()

            start_time = perf_counter()
            timer = Timer(250, handler.callback)
            try:
                self.assertTrue(timer.IsRunning())
                self.event_loop_helper.event_loop_until_condition(
                    handler.called_n(4),
                )
                self.assertTrue(timer.IsRunning())
            finally:
                timer.Stop()

            self.assertFalse(timer.IsRunning())
```

Then I ran into the situation mentioned in https://github.com/enthought/pyface/issues/516 where I get a timeout from `event_loop_until_condition`. Then if I increases the timeout tolerance on `event_loop_until_condition` to, say, 30.0:

```python
        for i in range(40):
            handler = ConditionHandler()

            start_time = perf_counter()
            timer = Timer(250, handler.callback)
            try:
                self.assertTrue(timer.IsRunning())
                self.event_loop_helper.event_loop_until_condition(
                    handler.called_n(4),
                    timeout=30.0,                            # <----- added this
                )
                self.assertTrue(timer.IsRunning())
            finally:
                timer.Stop()

            self.assertFalse(timer.IsRunning())

            time_intervals = [
                y - x for x, y in zip(handler.times[:-1], handler.times[1:])
            ]

            self.assertEqual(
                handler.count, 4,
                "Failed at iteration {}. Got time intervals: {}".format(
                    i, time_intervals,
                ))
```
Then I can reproduce the issue quite consistently:
```
======================================================================
FAIL: test_repeat (pyface.timer.tests.test_timer.TestTimer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/pyface/pyface/timer/tests/test_timer.py", line 447, in test_repeat
    i, time_intervals,
AssertionError: 5 != 4 : Failed at iteration 30. Got time intervals: [0.34967374200095946, 3.1618865769996773, 0.2882140330002585, 9.150152285999866]
```

The time interval is so much larger, which is consistent with the Qt bug mentioned in #516: https://bugreports.qt.io/browse/QTBUG-63832

Note that `event_loop_helper.event_loop_until_condition` uses a QTime to poll for the condition too, i.e. both the subject being tested and the testing tool may be suffering from the same bug.

#798 extended the assertion failure message, and on CI we can see this error on OSX + Qt:
```
======================================================================
FAIL: test_repeat (pyface.timer.tests.test_timer.TestTimer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/travis/.edm/envs/pyface-test-3.6-pyqt5/lib/python3.6/site-packages/pyface/timer/tests/test_timer.py", line 432, in test_repeat
    "Time intervals between calls: {}".format(time_intervals)
AssertionError: 5 != 4 : Time intervals between calls: [0.2964645700000119, 0.19584663000000546, 0.38052897699992627, 0.19247205300007408]
```
[pyface-sporadic.log](https://github.com/enthought/pyface/files/5531187/pyface-sporadic.log)


On my machine (OSX 10.15), after I managed to consistently reproduce the issue (see previous comment), I did the following changes to my machine:
```
$ defaults read NSGlobalDomain NSAppSleepDisabled
2020-11-12 14:24:27.075 defaults[11368:178841] 
The domain/default pair of (kCFPreferencesAnyApplication, NSAppSleepDisabled) does not exist
$ defaults write NSGlobalDomain NSAppSleepDisabled -bool YES
```
And then when I run the test, it passes every single time. So I am confident that it is AppNap getting involved, preventing Qt timer from being processed in order to save energy.

Then of course, one last sanity check, if I enable App Nap again, the test fails again:
```
$ defaults delete NSGlobalDomain NSAppSleepDisabled
(pyface-env) kchoi-mbp2019:pyface kchoi$ python -m unittest pyface.timer.tests.test_timer.TestTimer.test_repeat
F
======================================================================
FAIL: test_repeat (pyface.timer.tests.test_timer.TestTimer)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/pyface/pyface/timer/tests/test_timer.py", line 431, in test_repeat
    self.assertEqual(
AssertionError: 5 != 4 : Failed at iteration 32. Got time intervals: [7.549979257999993, 3.850039345000013, 0.25008981099999517, 10.050068199999998]

----------------------------------------------------------------------
Ran 1 test in 132.878s

FAILED (failures=1)
```

--- a/pyface/timer/tests/test_timer.py
+++ b/pyface/timer/tests/test_timer.py
@@ -422,9 +422,14 @@ class TestTimer(TestCase, GuiTestAssistant):
 
         self.assertFalse(timer.IsRunning())
 
-        self.assertEqual(handler.count, 4)
-
-        expected_times = [start_time + 0.2 * i + 0.2 for i in range(4)]
+        # The callback may be called more than 4 times depending
+        # on the order when condition timer in event_loop_unit_condition
+        # is called relative to the Timer here. Timer accuracy also depends
+        # on whether the event loop is interrupted by the system.
+        # The objective is that the timer should not be called too frequently.
+        expected_times = [
+            start_time + 0.2 * i + 0.2 for i in range(handler.count)
+        ]
 
         self.assertTrue(
             all(

